fairscale/optim/oss.py

```diff
@@ -83,21 +92,52 @@ def partition_parameters(self) -> List[List[dict]]:
         corresponds to rank 0, etc. We need all the ranks for the broadcast
         inside step().
         """
-        world_size = dist.get_world_size(self.group)
-        param_groups: List[List] = [list() for _ in range(world_size)]
-        sizes = [0] * world_size
-        for param_group in self.param_groups:
-            param_lists: List[List] = [list() for _ in range(world_size)]
-            for param in param_group["params"]:
-                # Add this param to rank with smallest size.
-                rank = sizes.index(min(sizes))
-                param_lists[rank].append(param)
-                sizes[rank] += param.numel()
-            for rank, params in enumerate(param_lists):
-                param_group_rank = copy.copy(param_group)
-                param_group_rank["params"] = params
-                param_groups[rank].append(param_group_rank)
-        return param_groups
+        if len(self._partition_parameters) == 0:
+            self._partition_parameters = [list() for _ in range(self.world_size)]
+            sizes = [0] * self.world_size
+            for param_group in self.param_groups:
+                param_lists: List[List] = [list() for _ in range(self.world_size)]
+                for param in param_group["params"]:
+                    # Add this param to rank with smallest size.
+                    rank = sizes.index(min(sizes))
+                    param_lists[rank].append(param)
+                    sizes[rank] += param.numel()
+
+                for rank, params in enumerate(param_lists):
+                    param_group_rank = copy.copy(param_group)
+                    param_group_rank["params"] = params
+                    self._partition_parameters[rank].append(param_group_rank)
+
+        return self._partition_parameters
+
+    @property
+    def per_device_params(self) -> List[List[Parameter]]:
```

Comment:
trying to keep all the partition related tidbits in the same place, and caching whenever possible